src/c/functions.c

```diff
@@ -34,6 +46,99 @@ void sum_vectors(const size_t size, const Vector *v1, const Vector *v2, Vector *
   }
 }
 
+/**
+ * Apply gravity to a vector of forces.
+ */
+void apply_gravity(const size_t size, const ParticleProperties *particles, Vector *forces) {
+  for (size_t i = 0; i < size; ++i) {
+    forces[i].y_component -= (particles[i].mass * 9.81d);
+  }
+}
+
+/**
+ * Compute the forces applied to P2 given it was collided by P1.
+ * Note: previous_normal and previous_tangent correspond to P2 with respect to P1.
+ */
+void collide_two_particles(const double dt, const Particle *p1, const Particle *p2, const Vector *velocity_p1, const Vector *velocity_p2, const ParticleProperties *properties_p2, Vector *force_p2, double *previous_normal, double *previous_tangent) {
+  const double distance = compute_distance(p1, p2);
+
+  const Vector normal = {
+    .x_component = (p2->x_coordinate - p1->x_coordinate) / distance,
+    .y_component = (p2->y_coordinate - p1->y_coordinate) / distance
+  };
+
+  const double normal_velocity = (normal.x_component * (velocity_p1->x_component - velocity_p2->x_component))
+      + (normal.y_component * (velocity_p1->y_component - velocity_p2->y_component));
+
+  const double tangent_velocity = (normal.y_component * (velocity_p1->x_component - velocity_p2->x_component))
+      - (normal.x_component * (velocity_p1->y_component - velocity_p2->y_component));
+
+  const double dfn = normal_velocity * properties_p2->kn * dt;
+  const double dfs = tangent_velocity * properties_p2->ks * dt;
+
+  double Fn_1_2 = *previous_normal + dfn;
+  double Fs_1_2 = *previous_tangent + dfs;
+
+  if (Fn_1_2 < 0) {
+    Fn_1_2 = 0;
+    Fs_1_2 = 0;
+  }
+
+  const double Fs_1_2_max = Fn_1_2 * tan((30 * M_PI) / 180);
+
+  if (abs(Fs_1_2) > Fs_1_2_max) {
+    Fs_1_2 = (abs(Fs_1_2_max) * abs(Fs_1_2)) / Fs_1_2;
+  }
+
+  // Update the forces of p2.
+  force_p2->x_component -= (-normal.x_component * Fn_1_2) - (normal.y_component * Fs_1_2);
+  force_p2->y_component -= (-normal.y_component * Fn_1_2) + (normal.x_component * Fs_1_2);
```

Comment:
I really do not understand why `-=` fixes the problems.
In matlab I see a `sum`... however it is a **matrix** and they _transpose_ it before and after the _sum_... maybe that is what makes it had the opposite sign.

I am not 100 percent happy with this, but at least it is working for now.
But, we certainly had to check what is happening here.